src/connectWithShell.tsx

```diff
@@ -143,3 +147,96 @@ export function connectWithShell<S = {}, OP = {}, SP = {}, DP = {}>(
         return wrapWithShellContext(component, mapStateToProps, mapDispatchToProps, boundShell, options)
     }
 }
+
+export interface ObservablesMap {
+    [key: string]: ObservableState<any>
+}
+
+export type ObservedSelectorsMap<M> = {
+    [K in keyof M]: M[K] extends ObservableState<infer S> ? S : undefined
+}
+
+export type OmitObservedSelectors<T, M> = Omit<T, keyof M>
+
+export function mapObservablesToSelectors<M extends ObservablesMap>(map: M): ObservedSelectorsMap<M> {
+    const result: Partial<ObservedSelectorsMap<M>> = {}
+    for (const key in map) {
+        const selector = map[key].current()
+        result[key] = selector
+    }
+    return result as ObservedSelectorsMap<M>
+}
+
+export function observeWithShell<OM extends ObservablesMap, OP extends ObservedSelectorsMap<OM>>(
+    observables: OM,
+    boundShell: Shell
+): <S, SP, DP>(
+    innerFactory: ConnectedComponentFactory<S, OP, SP, DP>
+) => ConnectedComponentFactory<S, OmitObservedSelectors<OP, OM>, SP, DP, OP> {
+    return <S, SP, DP>(innerFactory: ConnectedComponentFactory<S, OP, SP, DP>) => {
+        // exclude observed selectors from wrapper props, because we want those selectors to be in the wrapper's state instead
+        type ObservableWrapperProps = OmitObservedSelectors<OP, OM>
+        type ObservableWrapperState = ObservedSelectorsMap<OM>
+
+        const observableConnectedComponentFactory: ConnectedComponentFactory<S, ObservableWrapperProps, SP, DP, OP> = pureComponent => {
+            class ObservableWrapperComponent extends React.Component<ObservableWrapperProps, ObservableWrapperState> {
+                public connectedComponent: React.ComponentType<OP>
+                public unsubscribes: StateObserverUnsubscribe[]
+
+                constructor(props: OP) {
+                    super(props)
+                    this.connectedComponent = innerFactory(pureComponent)
+                    this.unsubscribes = []
+                    this.state = mapObservablesToSelectors(observables)
+                }
+
+                public componentDidMount() {
+                    for (const key in observables) {
+                        const unsubscribe = observables[key].subscribe(boundShell, () => {
+                            const oldState = this.state
+                            const newState = mapObservablesToSelectors(observables)
+                            if (newState !== oldState) {
+                                this.setState(newState)
+                            }
```

Comment:
The condition will always be false, due to equality by reference, setState is always called.
Better use `shouldComponentUpdate` / `_.isEqual` instead

Reply:
We get new object here when the state changes. This is how redux works by default. No need for other kinds of comparison here.